feat: MI Assign: Specify Activities & Respondents (M2-7135, M2-7507)

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

- [x] Tests for the changes have been added

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-7135](https://mindlogger.atlassian.net/browse/M2-7135)
🔗 [Jira Ticket M2-7507](https://mindlogger.atlassian.net/browse/M2-7507)

This chunky PR adds the first pane of Multi-Informant Activity Assignment, including:
- Creation of `ActivityAssignDrawer` component, based on MUI Drawer. This drawer can be opened from:
  - Applet > Activities screen "Assign" button
  - Applet > Activities > Activity/Activity Flow summary card's "Assign Activity" menu item
  - Applet > Participants > Participant's "Assign Activity" menu item
  - Applet > Participants > Participant Details screen "Assign Activity" button
  - Applet > Participants > Participant Details > Activities > Activity/Activity Flow summary card's "Assign Activity" menu item
  - Applet > Participants > Participant Details > Activity/Activity Flow DataViz screen "Assign" button
- Refactoring Take Now's participant dropdowns, so that they could be decoupled from Take Now-specific functionality. This dropdown component was previously known as `LabeledUserDropdown`, and has now been turned in a shared component called `ParticipantDropdown` and its associated hook, `useParticipantDropdown`. Refactoring took place over the course of several commits, each step explained in associated commit messages.
- Ensuring `ParticipantSnippet` is used consistently throughout the Admin App. This component displays a brief rich-text summary of a participant, and previously was rendering different data in Take Now, Applet Overview, and elsewhere. This in turn fixed [M2-7507](https://mindlogger.atlassian.net/browse/M2-7507).

> [!NOTE]
> The Search Bar was broken off into a [separate task](https://mindlogger.atlassian.net/browse/M2-7404), since it's not essential for functionality and doing so helped keep this PR's scope manageable.
> <img width="912" alt="be608966-2ed9-47a6-a2ca-be0ea93385e7" src="https://github.com/user-attachments/assets/2c1c6dc0-4d1e-451c-bfc0-81d9b71033ff">

> [!TIP]
> Commits are composed of related logic, though there are many of them. It's probably helpful to review [commit-by-commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1870/commits) to make it easier to understand the code and the process of building out this screen!

### 📸 Screenshots


https://github.com/user-attachments/assets/207dfa1b-1dbb-424b-a91d-2118eb851b82



### 🪤 Peer Testing

> [!NOTE] 
> This feature depends on the `enableActivityAssign` feature flag. This _should_ already be enabled in dev/local environments, but if needed, you can add `features.enableActivityAssign = true` prior to the `return` statement [here in `useFeatureFlags`](https://github.com/ChildMindInstitute/mindlogger-admin/blob/feature/M2-7135-mi-assign-specify-activities-respondents/src/shared/hooks/useFeatureFlags.ts#L25-L26).

Please refer to the [AC of the JIRA task](https://mindlogger.atlassian.net/browse/M2-7135) for an overview of what to test on this screen.


[M2-7507]: https://mindlogger.atlassian.net/browse/M2-7507?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

farmerpaul committed:
feat: Add ActivityAssignDrawer + event handlers

farmerpaul committed:
feat: add HelpPopup

Also create reusable styled component `BorderedOverflow` that renders a
vertically scrollable container that displays a light grey top/bottom
border whenever overflowing content exists beyond that edge.

Update universal overlay opacity to what's shown in Figma.

Fix font weight of default submit button of modals; it is never
indicated in Figma anywhere to override default button variant styling
for modals. Probably a relic of outdated styling.

farmerpaul committed:
feat: add `ActiviesList` & `ActivityCheckbox`

`ActivitesList` renders the scrollable list of avaialble activities and
activity flows, allowing the user to click rows to select/deselect
items. Required adding support for "size" prop to `Chip`s.

`ActivityCheckbox` is used in a couple places during activity selection,
so it made sense to create a reusable component. Required a new SVG
sprite, as well as applying `currentColor` to existing sprite fill
during use instead of in definition, due to the way SVG fill is
overridden using CSS in `Button` components.

farmerpaul committed:
refactor: add `hydrateActivityFlows` utility

There were already 2 instances where this is happening in the Dashboard,
and with `ActivityAssignDrawer` we will have a 3rd, so it made sense to
make a shared utility. This utility populates Activity Flows with their
activity entities (rather than just `activityIds`), which is needed to
render their thumbnails.

farmerpaul committed:
fix: use correct prop for respondent

Activity Assign is about respondents (user IDs) not source subject IDs,
so fixed that prop name and associated usage.

farmerpaul committed:
feat: add drawer pane infra & use `ActivitiesList`

Add lots of infra to `ActivityAssignDrawer`:
- form state management
- step management (selection vs. review) and transitions between steps
- footer button transitions
- populating `ActivitiesList` from the API
- selecting/deselecting activities

A note about how to maintain fade transitions between steps is added as
a comment. This approach is probably the simplest way to achieve fades,
but relies on one of the panes having a guaranteed taller content height
than the other. If this approach is deemed too fragile, that can be
addressed when M2-7261 is implemented.

Added some global theme styles that seemed appropriate. Worth noting is
the introduction of the `outlined` style variant for `IconButton`s,
which has been in the designs since ML1 but never implemented. Since
`IconButton`s lack a `variant` prop, used the `color` prop and a
custom-named color variant.

farmerpaul committed:
feat: add banner support to Activity Assign

Adapt existing global Banners infrastructure to `ActivityAssignDrawer`,
introducing a hook to make it convenient to manage display of local
banners without overcomplicating the component.

Also fix a layout bug in the Banner component (they were never being
rendered properly when there was no close button). Add support to
customize the icon as well.

Also fix a styling/layout issue with the footer reveal.

farmerpaul committed:
refactor: LabeledUserDropdown ⇒ ParticipantDropdown

Refactoring this to prepare it as a shared component. Step 1: rename it
to something that better describes its purpose.

farmerpaul committed:
refactor: extract ParticipantDropdown functions into hook

Move functions, utilities, and types specific to `ParticipantDropdown`s
into own hook, which can be used more portably rather than strictly in
the Take Now modal.

farmerpaul committed:
refactor: `ParticipantDropdown` ⇒ shared location

farmerpaul committed:
refactor: label/tooltip/warning ⇒ `TakeNowDropdown`

Extract more content into `TakeNowModal` component by creating
`TakeNowDropdown`, a Take Now-specific wrapper for `ParticipantDropdown`

farmerpaul committed:
refactor: respondent subject ID instead of user ID

It will be easier to manage assignments if we're using the same unique
ID across both respondent and target subject dropdowns, so switch to
subject ID instead of user ID.

farmerpaul committed:
refactor: add `isTeamMember` prop to options

Use new prop `isTeamMember` to identify when a dropdown option is part
of a team. This prop is initialized on creation based on the same logic
as before – checking whether the tag is "Team". This will enable us to
render the "Self" dropdown option without a "Team" tag beside it
when that "self" subject is a team member (as per designs), but still
retain its position in the "Team" grouping of dropdowns.

farmerpaul committed:
feat: add `AssignmentsTable` & validation

Add (partly unstyled) `AssignmentsTable` component to Activity Assign,
which lists activity assignments using `ParticipantDropdown` (wrapped
by `AssignmentDropdown`). Changes include:

- display of # of self-report & multi-informant assignments
- update form completion state that takes into account assignment rows
- `onSubmit` validation, which now checks for duplicate assignment rows
- add ability to style invalid rows in error state, by adding a new
  special prop to the `Row` type of both `DashboardTable` and `Table`
- optional "empty value" state for `ParticipantDropdown`, when a value
  is unselected and the user needs to be alerted
- replace matching `ParticipantDropdownOption` in target subject
  dropdown with "Self" when a respondent has been selected, and ensure
  `ParticipantDropdown` reacts to changes in the `options` prop to
  reflect this change
- update code comments for improved clarity in `ActivityAssignDrawer`

farmerpaul committed:
fix: improve UX of `ParticipantDropdown`

Ran into various UX issues in `ParticipantDropdown` when doing a lot
of rapid searching, focusing, and blurring of the component. Attempted
to fix most of these using some more nuanced focus state management.
Felt more natural to clear the dropdown when it loses focus so enabled
that by default. Tested to confirm these changes were appropriate for
Take Now modal as well.

farmerpaul committed:
fix: improve incomplete assignments validation

Pending some final recommendations from design, updated assignments
validation to allow the user to proceed to Review step if there are
empty assignment rows. This allows the user to clear duplicate
assignments, without a "delete row" button (currently not present in
designs). Empty rows are deleted when validation succeeds, as per
suggestion made in [this Figma thread](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x?node-id=9404-158430#881972698).

farmerpaul committed:
style: `ParticipantDropdown` design alignment (#1)

Step 1 of better aligning these dropdowns to Figma:

Display `ParticipantSnippet` when the component is unfocused. Fix
size of dropdown & clear icons, and padding of the text field and
snippet to accommodate.

farmerpaul committed:
style: `ParticipantDropdown` design alignment (#2)

Step 2 of better aligning these dropdowns to Figma:

Add new "full" variant to ParticipantDropdown for varied styling in
AssignmentsTable. Required adding hover effect and a number of spacing
adjustments.

farmerpaul committed:
fix: consolidate Team Member snippet render logic

When passing a Team Member to a `ParticipantSnippet`, there was logic
that only existed in `ParticipantDropdown` to render the right fields in
the right order. Moved that logic into the snippet component so that
Team Members are always rendered in the same way wherever the snippet is
used.

Also rename `ParticipantDropdown` i18n keys to account for its
independence from Take Now.

farmerpaul committed:
fix: update dropdown populating logic to follow AC

After re-reading the task's AC, discovered that the right dropdowns
should exclude Team Members, so updated that. Also, the Self option now
appears in its own unlabelled group at the top, regardless of where it
may have occurred naturally in the results, to follow designs.

Secondly, only the "Add Respondent" error state is present in the
designs (no "Add Subject", so removed that), and it's only shown if the
selected subject is a limited account. There's also more nuance around
whether to populate the right column with "self" based on left column
selection.

farmerpaul committed:
style: add Limited Account icon to dropdowns

farmerpaul committed:
revert: workaround for Banner layout transition

Timeout approach to properly reposition dropdown Paper after Banner
collapses (when clearing a duplicate row) doesn't actually work, so
reverted it for now.

farmerpaul committed:
fix: changes based on latest design feedback

Handle cases where both activity and participant are known when opening
the `ActivityAssignDrawer`, based on recent feedback from design. Also
instead of "invitations" use "emails" in assignment submission language.

farmerpaul committed:
test: fix broken tests

Participants screen, which now includes ActivityAssignDrawer, needs
some extra HTTP requests handled for their tests to continue passing.

farmerpaul committed:
test: add `ActivitesList` tests

Also use MUI Checkbox component for `ActivityCheckbox` for better a11y.

farmerpaul committed:
test: add tests for AssignmentsTable

In adding tests, fix some other issues:
- add aria-attribute to invalid table rows for a11y
- make correction to logic of conditionally issued API requests during
  search; issues with it were revealed during simultaneous test runs
- eliminate redundant dropdown search requests when value is changed
  programmatically
- fix styling

farmerpaul committed:
test: add tests for `HelpPopup`

farmerpaul committed:
test: add tests for `ActivityAssignDrawer`

In adding tests, discovered that when calling Activity Assign from the
context of a Team Member participant, the default assignment wasn't
quite correct (there should be no target subject selected), and we were
also missing alert text for this condition. So added support for these
scenarios.

farmerpaul committed:
fix: ParticipantSnippet consistency (M2-7507)

Update Applet Overview to align with how `ParticipantSnippet`s are
rendered elsewhere in the app. This small change was included in this PR
because it relates to making `ParticipantSnippet` usage consistent,
which was an essential part of the work in this PR.

farmerpaul committed:
test: add `ActivityAssignDrawer` submission tests

Also separate shared testing utilities into a `.test-utils.ts` file.

ChaconC commented on src/modules/Dashboard/components/ActivityAssignDrawer/ActivityAssignDrawer.tsx:
```diff
@@ -17,33 +42,217 @@ export const ActivityAssignDrawer = ({
+  // Reinitialize drawer form state whenever revealed
+  useEffect(() => {
+    if (open) {
+      reset(defaultValues);
```
Should reset `step`, as you get taken to the 2nd step if you close the drawer.

![CleanShot 2024-08-06 at 16 04 46](https://github.com/user-attachments/assets/cf23c651-0373-41c5-aa0c-941e5cd9eddc)

ChaconC commented on src/modules/Dashboard/components/ActivityAssignDrawer/ActivityAssignDrawer.tsx:
```diff
@@ -245,19 +284,56 @@ export const ActivityAssignDrawer = ({
+                name="assignments"
+                render={({ field: { onChange, value }, fieldState: { error } }) => {
+                  const duplicateRows =
+                    errors.assignments?.type === 'DuplicateRowsError'
```
This error is only validated after the first time you hit 'Submit'. Is it possible to trigger validation from the start?
| Before | After |
| - | - |
| ![CleanShot 2024-08-06 at 16 47 53](https://github.com/user-attachments/assets/256daa59-267e-469a-a715-d8df536bb4aa) | ![CleanShot 2024-08-06 at 16 47 42](https://github.com/user-attachments/assets/0849db2c-384b-4f6b-a7d0-69d9fb5b7a40) |

farmerpaul replied:
It would be possible to have validation behave that way, but the current behaviour was part of the AC, as derived from [Figma](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=9404-155473&t=S43bBddJBGbySLgw-4):

![image](https://github.com/user-attachments/assets/28b97b27-5322-402c-9a9d-96753f89865e)

ChaconC commented on src/modules/Dashboard/components/ActivityAssignDrawer/AssignmentsTable/AssignmentsTable.tsx:
```diff
@@ -0,0 +1,217 @@
+            ) {
+              updatedAssignments[index].targetSubjectId = respondentSubjectId;
+            }
+          }
```
This doesn't affect functionality and is more of a note on your comment, it's also not in any of the AC so feel free to ignore.

From the comment, it sounds like `Self` should only be auto selected if the previous assignment was `Self`, but there's no logic comparing to previous index and adding a new assignment always results in `Self` (**except** when you select a target first). 

https://github.com/user-attachments/assets/8ab9bde8-8aee-4b70-9bf6-b7a8e760f40b

farmerpaul replied:
Oh, in that comment, "previous assignment" is referring to the previously selected (or _currently_ selected) value in the dropdown of _that same row_ – not the assignment that is shown in the _previous row index_ of the table. I see how that's ambiguous. I'll update the comment to clarify.

Correct that this behaviour wasn't explicitly part of the AC, but it felt like a natural choice to me for "Self" to remain selected in the right column when the respondent changes in the left column, even though persisting "Self" in the subject column technically means you're changing the target subject ID in this instance.

ChaconC commented on src/modules/Dashboard/components/ParticipantDropdown/ParticipantDropdown.hooks.ts:
```diff
@@ -0,0 +1,271 @@
+  Roles.Manager,
+] as const;
+
+export const useParticipantDropdown = ({
```
Unrelated to this task, but showing "all participants" displays anonymous participants as targets for Assign (both target and respondent subjects)...which should probably not be possible.

![CleanShot 2024-08-06 at 16 19 23](https://github.com/user-attachments/assets/20873108-e272-46d7-8606-f974146c0c63)

farmerpaul replied:
Agreed, great catch. I don't think this should be selectable in Take Now, either, so I excluded those from the `useParticipantDropdown` hook level.

ChaconC reviewed (commented):
Amazing as always @farmerpaul, tested multiple scenarios and I think all AC is covered. Couple of Q's but I think we should be good to approve

farmerpaul committed:
fix: exclude anonymous respondents

farmerpaul committed:
fix: reset `step` when drawer is reopened

farmerpaul committed:
fix: clarify comment

farmerpaul committed:
fix: update copy & Activity Flow thumbnail styling

No longer using Participant name in pre-population alerts, so was able
to simplify banner components. Also fixed a small CSS bug with how
Activity Flow thumbnails were displaying.

farmerpaul committed:
test: fix tests related to prepopulation alerts

farmerpaul committed:
fix: restrict backdrop styling to Drawer
